### DISCUSSION TITLE
branching over sf problem after updating vctrs package

### DISCUSSION THREAD

**Discussion by hansvancalster** [ropensci/targets#1073] [Help] [answered]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

Hi,

We have an Rproj with a targets pipeline and the project's packages are maintained via renv. We recently updated from R 4.2 to R 4.3 and also updated the packages in our `renv.lock` file.

This has had some side-effects which we would like to share here.

In this process, the `vctrs` package was updated from 0.4.2 to 0.6.2.
This in turn affected how `tar_read()` aggregated over branches that contained sf objects (iteration = vector).
The internal targets function that got called was `targets:::value_produce_aggregate.tar_vector`, which uses `vctrs::vec_c` to aggregate over the sf pattern.
This worked fine with `vctrs` v0.4.2, but errored with 0.6.2. There are two problems with this and they are documented in two issues at the `vctrs` repo. See  https://github.com/r-lib/vctrs/issues/1839 and https://github.com/r-lib/vctrs/issues/1741.

Our workaround is to overwrite the targets internal function, before calling `tar_read`:

````
new_tar_vector <- function(value, objects) {
    out <- do.call(vctrs::vec_rbind, objects)
    if (any(sapply(out, inherits, what = "sfc"))) {
      sf::st_as_sf(out)
    } else {out}
}
try(assignInNamespace(
  "value_produce_aggregate.tar_vector",
  new_tar_vector,
  ns = "targets"))
````

This works, but is a nasty hack. We hope the issues get solved in the `vctrs` package, but if people read this and have other ideas of how to solve this, please let us know.

**Comment by wlandau** [comment, marked as answer]
Thanks for sharing and for following up with this issue in `vctrs`. The best I can advise is to use `iteration = "list"` instead of the default `iteration = "vector"` in `tar_target()` for `sf` objects.

**Comment by philiporlando** [comment]
@hansvancalster, thanks for the helpful write up on this! 

@wlandau, thanks for the `iteration` tip!

I'm also seeing this error after upgrading from `R-4.1.3` to `R-4.3.1`. Switching to `iteration = "list"` seems to be a decent workaround while we wait for the issue in `vctrs` to be resolved. I thought I'd add my reprex to this discussion instead of creating a new post. 

```r
#_targets.R
library(targets)


tar_option_set(
  packages = c(
    "dplyr",
    "sf"
    )
)


create_test_data <- function() {
 
  df <- data.frame(
    id = 1:3,
    state = c("CA", "TX", "NY")
  )

  points <- st_sfc(
    st_point(c(-119.4179, 36.7783)),  # CA
    st_point(c(-99.9018, 31.9686)),   # TX
    st_point(c(-74.0060, 40.7128))    # NY
  )
  
  df_sf <- st_sf(df, geometry = points)
   
}

subset_by_state <- function(df_sf, state) {
  
  df_sf %>% 
    dplyr::filter(state == {{ state }})
  
}


list(
  tar_target(
    name = data,
    command = create_test_data()
  ),
  tar_target(
    name = states,
    command = c("CA", "NY")
  ),
  tar_target(
    name = data_by_state,
    command = subset_by_state(data, state = states),
    pattern = map(states)
  )
)
```

```r
# interactive session
tar_make()

tar_read()
# <error/vctrs_error_scalar_type>
#   Error:
#   ! Input must be a vector, not a <sfc_POINT/sfc> object.
# Run `rlang::last_trace()` to see where the error occurred.

rlang::last_trace(drop = FALSE)
# <error/vctrs_error_scalar_type>
#   Error:
#   ! Input must be a vector, not a <sfc_POINT/sfc> object.
# ---
#   Backtrace:
#   ▆
# 1. ├─targets::tar_read(data_by_state)
# 2. │ └─targets::tar_read_raw(...) at targets/R/tar_read.R:37:2
# 3. │   └─targets:::tar_read_inner(name, branches, meta, path_store = store) at targets/R/tar_read_raw.R:37:2
# 4. │     ├─targets:::if_any(...) at targets/R/tar_read_raw.R:47:2
# 5. │     └─targets:::read_pattern(name, record, meta, branches, path_store) at targets/R/tar_read_raw.R:47:2
# 6. │       ├─targets:::value_produce_aggregate(value, objects) at targets/R/tar_read_raw.R:78:2
# 7. │       └─targets:::value_produce_aggregate.tar_vector(value, objects) at targets/R/class_value.R:36:2
# 8. │         ├─base::do.call(vctrs::vec_c, objects) at targets/R/class_vector.R:18:2
# 9. │         └─vctrs (local) `<fn>`(...)
# 10. └─vctrs:::stop_scalar_type(`<fn>`(`<sf_POINT>`), "", NULL)
# 11.   └─vctrs:::stop_vctrs(...)
# 12.     └─rlang::abort(message, class = c(class, "vctrs_error"), ..., call = call)
```
